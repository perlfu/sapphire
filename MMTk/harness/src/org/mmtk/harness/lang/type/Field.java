/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.mmtk.harness.lang.type;

/**
 * A field of a user defined object
 */
public class Field {
  /** Name of the field */
  private final String name;
  /** Type of the field */
  private final Type type;
  /** Offset of the field within the object's reference/int fields */
  private final int offset;

  /**
   * Create a new field
   * @param name
   * @param type
   */
  Field(String name, Type type, int offset) {
    this.name = name;
    this.type = type;
    this.offset = offset;
  }

  /**
   * @return The name of this field
   */
  public String getName() {
    return name;
  }

  /**
   * @return The type of the field
   */
  public Type getType() {
    return type;
  }

  /**
   * @return The word-offset to this field
   */
  public int getOffset() {
    return offset;
  }
}
