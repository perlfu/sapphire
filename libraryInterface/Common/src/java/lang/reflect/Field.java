/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang.reflect;

import java.lang.annotation.Annotation;

import org.jikesrvm.classloader.VM_Class;
import org.jikesrvm.classloader.VM_Field;
import org.jikesrvm.classloader.VM_TypeReference;
import org.jikesrvm.classloader.VM_Type;
import org.jikesrvm.classloader.VM_Atom;

import org.jikesrvm.objectmodel.VM_ObjectModel;
import org.jikesrvm.runtime.VM_Runtime;

/**
 * Implementation of java.lang.reflect.Field for JikesRVM.
 *
 * By convention, order methods in the same order
 * as they appear in the method summary list of Sun's 1.4 Javadoc API.
 */
public final class Field extends AccessibleObject implements Member {

  final VM_Field field;

  // Prevent this class from being instantiated.
  private Field() {
    field = null;
  }

  // For use by JikesRVMSupport
  Field(VM_Field f) {
    field = f;
  }

  public boolean equals(Object object) {
    if (object instanceof Field) {
      return field == ((Field)object).field;
    } else {
      return false;
    }
  }

  public Object get(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);

    VM_TypeReference type = field.getType();
    if (type.isReferenceType()) {
      return field.getObjectValueUnchecked(object);
    } else if (type.isCharType()) {
      return field.getCharValueUnchecked(object);
    } else if (type.isDoubleType()) {
      return field.getDoubleValueUnchecked(object);
    } else if (type.isFloatType()) {
      return field.getFloatValueUnchecked(object);
    } else if (type.isLongType()) {
      return field.getLongValueUnchecked(object);
    } else if (type.isIntType()) {
      return field.getIntValueUnchecked(object);
    } else if (type.isShortType()) {
      return field.getShortValueUnchecked(object);
    } else if (type.isByteType()) {
      return field.getByteValueUnchecked(object);
    } else if (type.isBooleanType()) {
      return field.getBooleanValueUnchecked(object);
    } else {
      throw new InternalError("Huh?  Field of unknown primitive type "+type);
    }
  }

  public boolean getBoolean(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getBooleanInternal(object);
  }

  public byte getByte(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getByteInternal(object);
  }

  public char getChar(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getCharInternal(object);
  }

  public Class<?> getDeclaringClass() {
    return field.getDeclaringClass().getClassForType();
  }

  public double getDouble(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getDoubleInternal(object);
  }

  public float getFloat(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getFloatInternal(object);
  }

  public int getInt(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getIntInternal(object);
  }

  public long getLong(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getLongInternal(object);
  }

  public int getModifiers() {
    return field.getModifiers();
  }

  public String getName() {
    return field.getName().toString();
  }

  public short getShort(Object object) throws IllegalAccessException, IllegalArgumentException {
    checkReadAccess(object);
    return getShortInternal(object);
  }

  public Class<?> getType() {
    return field.getType().resolve().getClassForType();
  }

  public int hashCode() {
    int code1 = getName().hashCode();
    int code2 = field.getDeclaringClass().toString().hashCode();
    return code1 ^ code2;
  }

  public boolean isSynthetic() {
    return field.isSynthetic();
  }

  public boolean isEnumConstant() {
    return field.isEnumConstant();
  }

  public void set(Object object, Object value)
    throws IllegalAccessException, IllegalArgumentException     {
    checkWriteAccess(object);

    VM_TypeReference type = field.getType();
    if (type.isReferenceType()) {
      if (value != null) {
        VM_Type valueType = VM_ObjectModel.getObjectType(value);
        VM_Type fieldType;
        try {
          fieldType = type.resolve();
        } catch (NoClassDefFoundError e) {
          throw new IllegalArgumentException("field type mismatch");
        }
        if (fieldType != valueType
            && !VM_Runtime.isAssignableWith(fieldType, valueType))
        {
          throw new IllegalArgumentException("field type mismatch");
        }
      }
      field.setObjectValueUnchecked(object, value);
    } else if (value instanceof Character) {
      setCharInternal(object, (Character) value);
    } else if (value instanceof Double) {
      setDoubleInternal(object, (Double) value);
    } else if (value instanceof Float) {
      setFloatInternal(object, (Float) value);
    } else if (value instanceof Long) {
      setLongInternal(object, (Long) value);
    } else if (value instanceof Integer) {
      setIntInternal(object, (Integer) value);
    } else if (value instanceof Short) {
      setShortInternal(object, (Short) value);
    } else if (value instanceof Byte) {
      setByteInternal(object, (Byte) value);
    } else if (value instanceof Boolean) {
      setBooleanInternal(object, (Boolean) value);
    } else {
      throw new IllegalArgumentException("field type mismatch");
    }
  }

  public void setBoolean(Object object, boolean value)
    throws IllegalAccessException, IllegalArgumentException    {
    checkWriteAccess(object);
    setBooleanInternal(object, value);
  }

  public void setByte(Object object, byte value)
    throws IllegalAccessException, IllegalArgumentException    {
    checkWriteAccess(object);
    setByteInternal(object, value);
  }

  public void setChar(Object object, char value)
    throws IllegalAccessException, IllegalArgumentException    {
    checkWriteAccess(object);
    setCharInternal(object, value);
  }

  public void setDouble(Object object, double value)
    throws IllegalAccessException, IllegalArgumentException    {
    checkWriteAccess(object);
    setDoubleInternal(object, value);
  }

  public void setFloat(Object object, float value)
    throws IllegalAccessException, IllegalArgumentException    {
    checkWriteAccess(object);
    setFloatInternal(object, value);
  }

  public void setInt(Object object, int value)
    throws IllegalAccessException, IllegalArgumentException    {
    checkWriteAccess(object);
    setIntInternal(object, value);
  }

  public void setLong(Object object, long value)
    throws IllegalAccessException, IllegalArgumentException    {
    checkWriteAccess(object);
    setLongInternal(object, value);
  }

  public void setShort(Object object, short value)
    throws IllegalAccessException, IllegalArgumentException   {
    checkWriteAccess(object);
    setShortInternal(object, value);
  }

  public String toString() {
    StringBuilder sb = new StringBuilder(64);
    Modifier.toString(getModifiers(), sb).append(' ');
    sb.append(JikesRVMHelpers.getUserName(getType())).append(' ');
    sb.append(getDeclaringClass().getName()).append('.');
    sb.append(getName());
    return sb.toString();
  }

  private void checkReadAccess(Object obj) throws IllegalAccessException,
                                                  IllegalArgumentException,
                                                  ExceptionInInitializerError {

    VM_Class declaringClass = field.getDeclaringClass();
    if (!field.isStatic()) {
      if (obj == null) {
        throw new NullPointerException();
      }

      VM_Type objType = VM_ObjectModel.getObjectType(obj);
      if (objType != declaringClass && !VM_Runtime.isAssignableWith(declaringClass, objType)) {
        throw new IllegalArgumentException();
      }
    }

    if (!field.isPublic() && !isAccessible()) {
      VM_Class accessingClass = VM_Class.getClassFromStackFrame(2);
      JikesRVMSupport.checkAccess(field, accessingClass);
    }

    if (field.isStatic() && !declaringClass.isInitialized()) {
      try {
        VM_Runtime.initializeClassForDynamicLink(declaringClass);
      } catch (Throwable e) {
        ExceptionInInitializerError ex = new ExceptionInInitializerError();
        ex.initCause(e);
        throw ex;
      }
    }
  }

  private void checkWriteAccess(Object obj) throws IllegalAccessException,
                                                   IllegalArgumentException,
                                                   ExceptionInInitializerError {

    VM_Class declaringClass = field.getDeclaringClass();
    if (!field.isStatic()) {
      if (obj == null) {
        throw new NullPointerException();
      }

      VM_Type objType = VM_ObjectModel.getObjectType(obj);
      if (objType != declaringClass && !VM_Runtime.isAssignableWith(declaringClass, objType)) {
        throw new IllegalArgumentException();
      }
    }

    if (!field.isPublic() && !isAccessible()) {
      VM_Class accessingClass = VM_Class.getClassFromStackFrame(2);
      JikesRVMSupport.checkAccess(field, accessingClass);
    }

    if (field.isFinal())
      throw new IllegalAccessException();

    if (field.isStatic() && !declaringClass.isInitialized()) {
      try {
        VM_Runtime.initializeClassForDynamicLink(declaringClass);
      } catch (Throwable e) {
        ExceptionInInitializerError ex = new ExceptionInInitializerError();
        ex.initCause(e);
        throw ex;
      }
    }
  }

  public boolean getBooleanInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (!type.isBooleanType()) throw new IllegalArgumentException("field type mismatch");
    return field.getBooleanValueUnchecked(object);
  }

  private byte getByteInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (!type.isByteType()) throw new IllegalArgumentException("field type mismatch");
    return field.getByteValueUnchecked(object);
  }

  private char getCharInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (!type.isCharType()) throw new IllegalArgumentException("field type mismatch");
    return field.getCharValueUnchecked(object);
  }

  private double getDoubleInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isDoubleType()) {
      return field.getDoubleValueUnchecked(object);
    } else if (type.isFloatType()) {
      return (double)field.getFloatValueUnchecked(object);
    } else if (type.isLongType()) {
      return (double)field.getLongValueUnchecked(object);
    } else if (type.isIntType()) {
      return (double)field.getIntValueUnchecked(object);
    } else if (type.isShortType()) {
      return (double)field.getShortValueUnchecked(object);
    } else if (type.isCharType()) {
      return (double)field.getCharValueUnchecked(object);
    } else if (type.isByteType()) {
      return (double)field.getByteValueUnchecked(object);
    } else {
      throw new IllegalArgumentException("field type mismatch");
    }
  }

  private float getFloatInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isFloatType()) {
      return field.getFloatValueUnchecked(object);
    } else if (type.isLongType()) {
      return (float)field.getLongValueUnchecked(object);
    } else if (type.isIntType()) {
      return (float)field.getIntValueUnchecked(object);
    } else if (type.isShortType()) {
      return (float)field.getShortValueUnchecked(object);
    } else if (type.isCharType()) {
      return (float)field.getCharValueUnchecked(object);
    } else if (type.isByteType()) {
      return (float)field.getByteValueUnchecked(object);
    } else {
      throw new IllegalArgumentException("field type mismatch");
    }
  }

  private int getIntInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isIntType()) {
      return field.getIntValueUnchecked(object);
    } else if (type.isShortType()) {
      return (int)field.getShortValueUnchecked(object);
    } else if (type.isCharType()) {
      return (int)field.getCharValueUnchecked(object);
    } else if (type.isByteType()) {
      return (int)field.getByteValueUnchecked(object);
    } else {
      throw new IllegalArgumentException("field type mismatch");
    }
  }

  private long getLongInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isLongType()) {
      return field.getLongValueUnchecked(object);
    } else if (type.isIntType()) {
      return (long)field.getIntValueUnchecked(object);
    } else if (type.isShortType()) {
      return (long)field.getShortValueUnchecked(object);
    } else if (type.isCharType()) {
      return (long)field.getCharValueUnchecked(object);
    } else if (type.isByteType()) {
      return (long)field.getByteValueUnchecked(object);
    } else {
      throw new IllegalArgumentException("field type mismatch");
    }
  }

  private short getShortInternal(Object object) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isShortType()) {
      return field.getShortValueUnchecked(object);
    } else if (type.isByteType()) {
      return (short)field.getByteValueUnchecked(object);
    } else {
      throw new IllegalArgumentException("field type mismatch");
    }
  }

  private void setBooleanInternal(Object object, boolean value)
    throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isBooleanType())
      field.setBooleanValueUnchecked(object, value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  private void setByteInternal(Object object, byte value) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isByteType())
      field.setByteValueUnchecked(object, value);
    else if (type.isLongType())
      field.setLongValueUnchecked(object, (long)value);
    else if (type.isIntType())
      field.setIntValueUnchecked(object, (int)value);
    else if (type.isShortType())
      field.setShortValueUnchecked(object, (short)value);
    else if (type.isCharType())
      field.setCharValueUnchecked(object, (char)value);
    else if (type.isDoubleType())
      field.setDoubleValueUnchecked(object, (double)value);
    else if (type.isFloatType())
      field.setFloatValueUnchecked(object, (float)value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  private void setCharInternal(Object object, char value) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isCharType())
      field.setCharValueUnchecked(object, value);
    else if (type.isLongType())
      field.setLongValueUnchecked(object, (long)value);
    else if (type.isIntType())
      field.setIntValueUnchecked(object, (int)value);
    else if (type.isShortType())
      field.setShortValueUnchecked(object, (short)value);
    else if (type.isDoubleType())
      field.setDoubleValueUnchecked(object, (double)value);
    else if (type.isFloatType())
      field.setFloatValueUnchecked(object, (float)value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  private void setDoubleInternal(Object object, double value) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isDoubleType())
      field.setDoubleValueUnchecked(object, value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  private void setFloatInternal(Object object, float value) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isFloatType())
      field.setFloatValueUnchecked(object, value);
    else if (type.isDoubleType())
      field.setDoubleValueUnchecked(object, (double)value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  private void setIntInternal(Object object, int value) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isIntType())
      field.setIntValueUnchecked(object, value);
    else if (type.isLongType())
      field.setLongValueUnchecked(object, (long) value);
    else if (type.isDoubleType())
      field.setDoubleValueUnchecked(object, (double)value);
    else if (type.isFloatType())
      field.setFloatValueUnchecked(object, (float)value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  private void setLongInternal(Object object, long value) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isLongType())
      field.setLongValueUnchecked(object, value);
    else if (type.isDoubleType())
      field.setDoubleValueUnchecked(object, (double)value);
    else if (type.isFloatType())
      field.setFloatValueUnchecked(object, (float)value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  private void setShortInternal(Object object, short value) throws IllegalArgumentException {
    VM_TypeReference type = field.getType();
    if (type.isShortType())
      field.setShortValueUnchecked(object, value);
    else if (type.isLongType())
      field.setLongValueUnchecked(object, (long)value);
    else if (type.isIntType())
      field.setIntValueUnchecked(object, (int)value);
    else if (type.isDoubleType())
      field.setDoubleValueUnchecked(object, (double)value);
    else if (type.isFloatType())
      field.setFloatValueUnchecked(object, (float)value);
    else
      throw new IllegalArgumentException("field type mismatch");
  }

  // AnnotatedElement interface

  public Annotation[] getDeclaredAnnotations() {
    return field.getDeclaredAnnotations();
  }

  public <T extends Annotation> T getAnnotation(Class<T> annotationClass) {
    return field.getAnnotation(annotationClass);
  }

  // Generics support

  public Type getGenericType() {
    VM_Atom signature = field.getSignature();
    if (signature == null) {
      return getType();
    } else {
      return JikesRVMHelpers.getFieldType(this, signature);
    }
  }

  public String toGenericString() {
    StringBuilder sb = new StringBuilder(64);
    Modifier.toString(getModifiers(), sb).append(' ');
    sb.append(getGenericType()).append(' ');
    sb.append(getDeclaringClass().getName()).append('.');
    sb.append(getName());
    return sb.toString();
  }
}
