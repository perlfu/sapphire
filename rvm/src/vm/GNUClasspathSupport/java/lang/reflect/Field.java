/*
 * (C) Copyright IBM Corp 2002
 */
//$Id$
package java.lang.reflect;

import com.ibm.JikesRVM.librarySupport.ReflectionSupport;
import com.ibm.JikesRVM.VM_Field;

/**
 * Library support interface of Jikes RVM
 *
 * @author Julian Dolby
 *
 */
public final class Field extends AccessibleObject implements Member {

    VM_Field field;

    /**
     * Prevent this class from being instantiated.
     */
    private Field() {}
    
    Field( VM_Field f ) {
	field = f;
    }
    
    public boolean equals(Object object) {
        if ( object == null) return false;
	if(this == object) return true;
	
	if(!(object instanceof Field)) return false;
	
        Field other = (Field) object;
        if ( field != null ) 
	    return field.equals( other.field );
        else 
	    return super.equals( object );
    }

    public Object get(Object object) throws IllegalAccessException, IllegalArgumentException
    {
	// TODO: check for Illegal Access Exception and Illegal Argument Exception
	
	if ((object == null) && (!field.isStatic()))  
	    throw new java.lang.NullPointerException();
	
	return (field.getObject(object));
    }

    public boolean getBoolean(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getBooleanValue( object );
    }
    

    public byte getByte(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getByteValue( object );
    }

    public char getChar(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getCharValue( object );
    }

    public Class getDeclaringClass() {
	return field.getDeclaringClass().getClassForType();
    }

    public double getDouble(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getDoubleValue( object );
    }

    public float getFloat(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getFloatValue( object );
    }

    public int getInt(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getIntValue( object );
    }

    public long getLong(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getLongValue( object );
    }

    public int getModifiers() {
	return field.getModifiers();
    }

    public String getName() {
	return field.getName().toString();
    }
    
    public short getShort(Object object) throws IllegalAccessException, IllegalArgumentException {
	return field.getShortValue( object );
    }

    public String getSignature() {
	return field.getDescriptor().toString();
    }

    public Class getType() {
	return field.getType().getClassForType();
    }

    public int hashCode() {
	return getName().hashCode();
    }

    public void set(Object object, Object value) throws IllegalAccessException, IllegalArgumentException {
	ReflectionSupport.setField(this,object,value);
    }

    public void setBoolean(Object object, boolean value) throws IllegalAccessException, IllegalArgumentException {
	field.setBooleanValue( object, value );
    }

    public void setByte(Object object, byte value) throws IllegalAccessException, IllegalArgumentException {
	field.setByteValue( object, value );
    }

    public void setChar(Object object, char value) throws IllegalAccessException, IllegalArgumentException {
	field.setCharValue( object, value );
    }

    public void setDouble(Object object, double value) throws IllegalAccessException, IllegalArgumentException {
	field.setDoubleValue( object, value );
    }

    public void setFloat(Object object, float value) throws IllegalAccessException, IllegalArgumentException {
	field.setFloatValue( object, value );
    }

    public void setInt(Object object, int value) throws IllegalAccessException, IllegalArgumentException {
	field.setIntValue( object, value );
    }

    public void setLong(Object object, long value) throws IllegalAccessException, IllegalArgumentException {
	field.setLongValue( object, value );
    }

    public void setShort(Object object, short value) throws IllegalAccessException, IllegalArgumentException {
	field.setShortValue( object, value );
    }

    public String toString() {
	StringBuffer buf;
	Class current;
	int arity = 0;

	buf = new StringBuffer();
	buf.append(Modifier.toString(getModifiers()));
	buf.append(" ");

	current = getType();
	while(current.isArray())
	{
		current = current.getComponentType();
		arity++;
	}
	buf.append(current.getName());
	for(;arity > 0; arity--) buf.append("[]");

	buf.append(" ");
	buf.append(getDeclaringClass().getName());
	buf.append(".");
	buf.append(getName());
	return buf.toString();
    }
}
