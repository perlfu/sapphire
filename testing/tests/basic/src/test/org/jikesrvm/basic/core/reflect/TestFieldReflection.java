/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.reflect;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.Comparator;

class TestFieldReflection {

  @SuppressWarnings({"CanBeFinal"})
  static class TypeWithPublicFields {
    public static boolean sboolean = true;
    public static byte sbyte = 127;
    public static short sshort = 1;
    public static int sint = 1;
    public static long slong = 1;
    public static float sfloat = 1.0F;
    public static double sdouble = 1.0;

    public boolean mboolean = true;
    public byte mbyte = 127;
    public short mshort = 1;
    public int mint = 1;
    public long mlong = 1;
    public float mfloat = 1.0F;
    public double mdouble = 1.0;
  }

  @SuppressWarnings({"CanBeFinal","UnusedDeclaration"})
  static class TypeWithDifferentAccessModifiers {
    public static boolean sboolean = true;
    protected static byte sbyte = 127;
    static short sshort = 1;
    private static int sint = 1;
    public static long slong = 1;
    protected static float sfloat = 1.0F;
    static double sdouble = 1.0;

    private boolean mboolean = true;
    byte mbyte = 127;
    protected short mshort = 1;
    public int mint = 1;
    protected long mlong = 1;
    float mfloat = 1.0F;
    private double mdouble = 1.0;
  }

  public static void main(String[] args) throws Exception {
    testFieldReflection(TypeWithPublicFields.class);
    testFieldReflection(TypeWithDifferentAccessModifiers.class);
  }

  private static void testFieldReflection(final Class<?> type) throws Exception {
    System.out.println("Testing on getFields");
    testFieldReflection(type.newInstance(), type, type.getFields());

    System.out.println("Testing on getDeclaredFields()");
    testFieldReflection(type.newInstance(), type, type.getDeclaredFields());
  }

  private static void testFieldReflection(final Object object, final Class<?> type, final Field[] fields) throws Exception {
    Arrays.sort(fields, StringComparator.COMPARATOR);

    printFields(object, fields);

    testBoolean(object, fields);
    printFields(object, fields);

    testByte(object, fields);
    printFields(object, fields);

    testShort(object, fields);
    printFields(object, fields);

    testInt(object, fields);
    printFields(object, fields);

    testLong(object, fields);
    printFields(object, fields);

    testFloat(object, fields);
    printFields(object, fields);

    testDouble(object, fields);
    printFields(object, fields);

    testType(type, "sboolean", boolean.class);
    testType(type, "sbyte", byte.class);
    testType(type, "sshort", short.class);
    testType(type, "sint", int.class);
    testType(type, "slong", long.class);
    testType(type, "sfloat", float.class);
    testType(type, "sdouble", double.class);
    testType(type, "mboolean", boolean.class);
    testType(type, "mbyte", byte.class);
    testType(type, "mshort", short.class);
    testType(type, "mint", int.class);
    testType(type, "mlong", long.class);
    testType(type, "mfloat", float.class);
    testType(type, "mdouble", double.class);
  }

  private static void testType(final Class<?> type, final String fieldName, final Class expected) {
    try {
      final Class<?> f = type.getDeclaredField(fieldName).getType();
      if (!expected.equals(f)) System.out.println(expected + " is not equal to " + f);
    } catch (NoSuchFieldException e) {
      System.out.println("testType: Unable to retrieve field " + fieldName + " from class " + type.getName());
    }
  }

  private static void printFields(final Object object, final Field[] fields) {
    System.out.print("Values:");
    for (final Field field : fields) {
      String value = "";
      try {
        value = String.valueOf(field.get(object));
      } catch (IllegalArgumentException e) {
        value = "!";
      } catch (IllegalAccessException e) {
        value = "-";
      }
      System.out.printf(" %5s",value);
    }
    System.out.println();
  }

  private static void testBoolean(final Object object, final Field[] fields) {
    banner("boolean", "false");
    for (final Field field : fields) {
      try {
        field.setBoolean(object, false);
        changeSuccess();
      } catch (Exception e) { changeError(e); }
    }
    System.out.println();
  }

  private static void testByte(final Object object, final Field[] fields) {
    banner("byte", "12");
    for (final Field field : fields) {
      try {
        field.setByte(object, (byte) 12);
        changeSuccess();
      } catch (Exception e) { changeError(e); }
    }
    System.out.println();
  }

  private static void testShort(final Object object, final Field[] fields) {
    banner("short", "2");
    for (final Field field : fields) {
      try {
        field.setShort(object, (short) 2);
        changeSuccess();
      } catch (Exception e) { changeError(e); }
    }
    System.out.println();
  }

  private static void testInt(final Object object, final Field[] fields) {
    banner("int", "3");
    for (final Field field : fields) {
      try {
        field.setInt(object, 3);
        changeSuccess();
      } catch (Exception e) { changeError(e); }
    }
    System.out.println();
  }

  private static void testLong(final Object object, final Field[] fields) {
    banner("long", "4");
    for (final Field field : fields) {
      try {
        field.setLong(object, 4);
        changeSuccess();
      } catch (Exception e) { changeError(e); }
    }
    System.out.println();
  }

  private static void testFloat(final Object object, Field[] fields) {
    banner("float", "5");
    for (final Field field : fields) {
      try {
        field.setFloat(object, 5.0F);
        changeSuccess();
      } catch (Exception e) { changeError(e); }
    }
    System.out.println();
  }

  private static void testDouble(final Object object, Field[] fields) {
    banner("double", "6.0");
    for (final Field field : fields) {
      try {
        field.setDouble(object, 6.0);
        changeSuccess();
      } catch (Exception e) { changeError(e); }
    }
    System.out.println();
  }

  private static void banner(final String typename, final Object value) {
    System.out.print("Set " + typename + "s to " + value + " ");
  }

  private static void changeSuccess() { System.out.print("+"); }

  private static void changeError(final Exception e) {
    if (e instanceof IllegalAccessException) {
      System.out.print("-");
    } else {
      System.out.print("!");
    }
  }

  private static class StringComparator implements Comparator<Object> {
    static StringComparator COMPARATOR = new StringComparator();

    public int compare(final Object x, final Object y) {
      return x.toString().compareTo(y.toString());
    }
  }
}

