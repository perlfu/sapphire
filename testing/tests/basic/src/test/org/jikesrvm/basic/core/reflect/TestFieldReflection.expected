Testing on getFields
Values:  true   127   1.0   1.0     1     1     1  true   127   1.0   1.0     1     1     1
Set booleans to false +!!!!!!+!!!!!!
Values: false   127   1.0   1.0     1     1     1 false   127   1.0   1.0     1     1     1
Set bytes to 12 !++++++!++++++
Values: false    12  12.0  12.0    12    12    12 false    12  12.0  12.0    12    12    12
Set shorts to 2 !!+++++!!+++++
Values: false    12   2.0   2.0     2     2     2 false    12   2.0   2.0     2     2     2
Set ints to 3 !!++++!!!++++!
Values: false    12   3.0   3.0     3     3     2 false    12   3.0   3.0     3     3     2
Set longs to 4 !!++!+!!!++!+!
Values: false    12   4.0   4.0     3     4     2 false    12   4.0   4.0     3     4     2
Set floats to 5 !!++!!!!!++!!!
Values: false    12   5.0   5.0     3     4     2 false    12   5.0   5.0     3     4     2
Set doubles to 6.0 !!+!!!!!!+!!!!
Values: false    12   6.0   5.0     3     4     2 false    12   6.0   5.0     3     4     2
Testing on getDeclaredFields()
Values:  true   127   1.0   1.0     1     1     1 false    12   6.0   5.0     3     4     2
Set booleans to false +!!!!!!+!!!!!!
Values: false   127   1.0   1.0     1     1     1 false    12   6.0   5.0     3     4     2
Set bytes to 12 !++++++!++++++
Values: false    12  12.0  12.0    12    12    12 false    12  12.0  12.0    12    12    12
Set shorts to 2 !!+++++!!+++++
Values: false    12   2.0   2.0     2     2     2 false    12   2.0   2.0     2     2     2
Set ints to 3 !!++++!!!++++!
Values: false    12   3.0   3.0     3     3     2 false    12   3.0   3.0     3     3     2
Set longs to 4 !!++!+!!!++!+!
Values: false    12   4.0   4.0     3     4     2 false    12   4.0   4.0     3     4     2
Set floats to 5 !!++!!!!!++!!!
Values: false    12   5.0   5.0     3     4     2 false    12   5.0   5.0     3     4     2
Set doubles to 6.0 !!+!!!!!!+!!!!
Values: false    12   6.0   5.0     3     4     2 false    12   6.0   5.0     3     4     2
Testing on getFields
Values:     1  true     1
Set booleans to false !+!
Values:     1 false     1
Set bytes to 12 +!+
Values:    12 false    12
Set shorts to 2 +!+
Values:     2 false     2
Set ints to 3 +!+
Values:     3 false     3
Set longs to 4 !!+
Values:     3 false     4
Set floats to 5 !!!
Values:     3 false     4
Set doubles to 6.0 !!!
Values:     3 false     4
Testing on getDeclaredFields()
Values:   127   1.0     -     -     -     1     1   127   1.0     1 false     4   1.0     1
Set booleans to false !!---!!!!!+!!!
Values:   127   1.0     -     -     -     1     1   127   1.0     1 false     4   1.0     1
Set bytes to 12 ++---+++++!+++
Values:    12  12.0     -     -     -    12    12    12  12.0    12 false    12  12.0    12
Set shorts to 2 !+---++!++!+++
Values:    12   2.0     -     -     -     2     2    12   2.0     2 false     2   2.0     2
Set ints to 3 !+---+!!++!++!
Values:    12   3.0     -     -     -     3     2    12   3.0     3 false     3   3.0     2
Set longs to 4 !+---+!!+!!++!
Values:    12   4.0     -     -     -     4     2    12   4.0     3 false     4   4.0     2
Set floats to 5 !+---!!!+!!!+!
Values:    12   5.0     -     -     -     4     2    12   5.0     3 false     4   5.0     2
Set doubles to 6.0 !!---!!!!!!!+!
Values:    12   5.0     -     -     -     4     2    12   5.0     3 false     4   6.0     2
